scalyr_agent/copying_manager/worker.py

```diff
@@ -1215,7 +1215,83 @@ def create_shared_object_manager(worker_class, worker_proxy_class):
     """
 
     class _SharedObjectManager(multiprocessing.managers.SyncManager):
-        pass
+        # the PID of the parent process.
+        pid = None
+        # When the shared object manager's process is going to be terminated,
+        # we need to be sure that the worker is stopped first.
+        # This class attribute should provide access to the worker for the watchdog that is implemented below.
+        worker = None
+
+        @classmethod
+        def _create_worker(cls, *args, **kwargs):
+            """
+            Create a new worker and save it as a 'cls.worker' class attribute
+            to be able to access the worker's instance by the watchdog.
+            :return: the proxy object for the 'cls.worker'
+            """
+            cls.worker = worker_class(*args, **kwargs)
+            return cls.worker
+
+        @classmethod
+        def _run_server(cls, *args, **kwargs):
+            """
+            This is the entry point of the SyncManager's process
+            and it is overridden to handle the situation where parent process was killed.
+            The current implementation of the multiprocess.managers does not provide ability to detect such situation
+            and the orphan process of the SyncManager still continues working even after the parent was killed.
+
+
+            To achieve needed behaviour, before we call the original '_run_server', we create
+            and start a 'watchdog' thread, which checks the existence of the parent process
+            and terminates everything after the parent dies.
+            :param args: passed directly to the original parent method.
+            :param kwargs: passed directly to the original parent method.
+            """
+
+            def run_watchdog():
+                """
+                Periodically check the parent process for existence and terminate everything if process is not found.
+                :return:
+                """
+                while True:
+                    try:
+                        # probing the parent process.
+                        if cls.pid is not None:
+                            os.kill(cls.pid, 0)
+                        time.sleep(1)
+                    except OSError:
+                        # parent is not found.
+                        # stop the worker if it is still alive.
+                        if cls.worker and cls.worker.is_alive():
+                            try:
```

Comment:
Should we also log something in this scenario? Does it even make sense? I assume it wouldn't make it up to scalyr until the next time the agent is started again, or?

Reply:
Sorry in advance if i haven't understood you correctly. The worker lives in the same process as this thread lives. When the parent probing is unsuccessful(parent is killed), then it's only needed to terminate/kill the current process and the worker thread will stop too. (But before that we try to stop it gracefully of course)

Reply:
Sorry, I just wanted to say it would be a good idea to log a message when watchdog kills a worker process.

So we have it for "audit" purposes and to potentially make troubleshooting easier (to know why a worker process has exited. etc.).

My only question is, if we log a message here, would it even make it to the Scalyr API?

Reply:
The logger of the current worker should be still ok even if we stop the worker. But the agent*.log file for this worker may be sent to Scalyr by a completely different worker, which may be also stopped at this moment.

Reply:
OK, in this case we should try to log it, I think.

Reply:
But even considering what I've said above, it still should be a good idea, because at least, those log messages can be read locally from the file. Thank for pointing on that.

Reply:
PS: Btw, in the normal case, the agent should be restarted right after the crash, so those agent*.log messages should reach the servers anyway.

Reply:
I agree about us wanting to log something.  BUT, I think we should only do that if the worker has already stopped or if when we ask it to stop, it gracefully does so.  I know you could argue that the second case should be logged, but I worry that due to race conditions, it might be normal for the worker to not have been told to stopped before we see the parent go away.  Hmm... maybe not? 

 @ArthurKamalov in the normal case, will the main agent process wait until all of the workers have terminated.  If so, then yes, let's always log.